Allow WAL volume to be specified at cluster create

cbandy opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the new behavior (if this is a feature change)?**

This PR allows one to specify a separate WAL volume to better control the performance of PostgreSQL.

Storage for WAL may be specified during cluster creation, and that spec carries through to replicas and clones. When not specified, WAL resides within the PGDATA directory and volume.

There is an operator configuration to specify WAL storage on all newly created clusters. When in use, the size of the volume can be overridden per-cluster.

This implementation does not define the WAL volume in any deployment templates because the volume name and mount path are constant.

Requires containers from https://github.com/CrunchyData/crunchy-containers/pull/1224.

cbandy commented on apiserver/root.go:
```diff
@@ -189,11 +189,6 @@ func initConfig() {
 	}
 	log.Infof("BasicAuth is %v", BasicAuth)
-
-	if !validStorageSettings() {
-		log.Error("Storage Settings are not defined correctly, can't continue")
-		os.Exit(2)
-	}
```
This validation is already taking place in `PgoConfig.GetConfig`.

cbandy commented on conf/postgres-operator/cluster-deployment.json:
```diff
@@ -157,9 +157,6 @@
                         },
                         {{ end }}
                         {
-                            "mountPath": "/pgwal",
-                            "name": "pgwal-volume"
```
This was present but never used. The container only touches `/pgwal` when `PGHA_WALDIR` env var is set.

cbandy commented on pgo/cmd/df.go:
```diff
@@ -57,9 +57,10 @@ const (
+	msgs.PVCTypePostgreSQL:    "data",
+	msgs.PVCTypepgBackRest:    "pgbackrest",
+	msgs.PVCTypeTablespace:    "tablespace",
+	msgs.PVCTypeWriteAheadLog: "wal",
```
This type appears like so:

```
PVC                 INSTANCE                       POD                                             TYPE        USED      CAPACITY  % USED
------------------- ------------------------------ ----------------------------------------------- ----------- --------- --------- ------
mycluster-pgbr-repo mycluster-backrest-shared-repo mycluster-backrest-shared-repo-6c79dbd644-zzf6h pgbackrest  11MiB     286MiB    4%
mycluster-sxkl      mycluster-sxkl                 mycluster-sxkl-65d787dd56-kggh8                 data        31MiB     286MiB    11%
mycluster-sxkl-wal  mycluster-sxkl                 mycluster-sxkl-65d787dd56-kggh8                 wal         64MiB     2GiB      3%
mycluster-wrjm      mycluster-wrjm                 mycluster-wrjm-69b56fd696-6w5l4                 data        31MiB     286MiB    11%
mycluster-wrjm-wal  mycluster-wrjm                 mycluster-wrjm-69b56fd696-6w5l4                 wal         144MiB    2GiB      7%
```

jkatz replied:
word

jkatz commented on apiservermsgs/dfmsgs.go:
```diff
@@ -26,6 +26,7 @@ const (
 	PVCTypePostgreSQL DfPVCType = iota
 	PVCTypepgBackRest
 	PVCTypeTablespace
+	PVCTypeWriteAheadLog
```
`PVCTypeWAL` would likely suffice. OK either ay.

jkatz commented on config/volumes.go:
```diff
@@ -15,10 +15,24 @@ package config
+
+// PostgreSQLWALPath returns the absolute path to a mounted WAL directory.
+func PostgreSQLWALPath(cluster string) string {
+	return PostgreSQLWALVolumeMount().MountPath + "/" + cluster + "-wal"
```
My preference would be to use a `fmt.Sprintf` here

jkatz commented on docs/content/Installation/operator-install.md:
```diff
@@ -120,8 +120,7 @@ Here are some common examples of configuration changes most installers would mak
     PrimaryStorage: gce
-    XlogStorage: gce
-    ArchiveStorage: gce
+    WALStorage: gce
```
Hm, but isn't this not actually defined? It appears to not be in the defaults that I've seen are empty.

It seems if we specify this, then one is automatically opted into using the WALStorage feature:

```go
		if apiserver.Pgo.WALStorage != "" {
			spec.WALStorage, _ = apiserver.Pgo.GetStorageSpec(apiserver.Pgo.WALStorage)
		}
```

cbandy replied:
> Hm, but isn't this not actually defined? It appears to not be in the defaults that I've seen are empty.

Yes? As committed, `WALStorage` is present and blank `pgo.yaml`.

> It seems if we specify this, then one is automatically opted into using the WALStorage feature:
> 
> ```go
> 		if apiserver.Pgo.WALStorage != "" {
> 			spec.WALStorage, _ = apiserver.Pgo.GetStorageSpec(apiserver.Pgo.WALStorage)
> 		}
> ```

Yes. This is why the shipped pgo.yaml configuration is blank. Setting it there applies to all newly created clusters.

When it is blank in pgo.yaml and the user wants a WAL volume, they specify a storage configuration during cluster create, similar to a tablespace.

jkatz commented on operator/backrest/restore.go:
```diff
@@ -143,21 +140,14 @@ func Restore(restclient *rest.RESTClient, namespace string, clientset *kubernete
-			return
+	// combine supplemental groups from all volumes
+	var supplementalGroups []int64
+	{
+		supplementalGroups = append(supplementalGroups, dataVolume.SupplementalGroups...)
+		for _, v := range tablespaceVolumes {
+			supplementalGroups = append(supplementalGroups, v.SupplementalGroups...)
 		}
 	}
```
Not totally sure how I feel about the outer `{ }` here

cbandy replied:
These brackets can be removed. In an earlier (unsubmitted) version, there was more going on inside.

jkatz commented on operator/backrest/restore.go:
```diff
@@ -349,6 +325,15 @@ func CreateRestoredDeployment(restclient *rest.RESTClient, cluster *crv1.Pgclust
 
+	// combine supplemental groups from all volumes
+	var supplementalGroups []int64
+	{
+		supplementalGroups = append(supplementalGroups, dataVolume.SupplementalGroups...)
+		for _, v := range tablespaceVolumes {
+			supplementalGroups = append(supplementalGroups, v.SupplementalGroups...)
+		}
+	}
```
Similarly, not sure how I feel about this grouping. Is there cited precedent for using it in this way (e.g. Effective go?)

cbandy replied:
Nothing in Effective Go. I've seen it used to control scope, often in main() or other "wiring" code, where some intermediate values (e.g. parsed config) are needed only briefly.

Nothing like that here.

jkatz commented on operator/cluster/clone.go:
```diff
@@ -238,9 +195,13 @@ func cloneStep1(clientset *kubernetes.Clientset, client *rest.RESTClient, namesp
-	// that sooner
-	createPVCs(clientset, client, task, namespace, sourcePgcluster, targetClusterName)
+	// create PVCs for pgBackRest and PostgreSQL
+	_, _, _, _, err = createPVCs(clientset, client, task, namespace, sourcePgcluster, targetClusterName)
+	if err != nil {
```
seems like a good oneliner place:

```go
if _, _, _, _, err := createPVCs(clientset, client, task, namespace, sourcePgcluster, targetClusterName); err != nil {
```

jkatz commented on operator/cluster/clone.go:
```diff
@@ -682,57 +670,62 @@ func createPgBackRestRepoSyncJob(clientset *kubernetes.Clientset, namespace stri
-	task *crv1.Pgtask, namespace string, sourcePgcluster crv1.Pgcluster, targetClusterName string) {
+	task *crv1.Pgtask, namespace string, sourcePgcluster crv1.Pgcluster, targetClusterName string,
+) (
+	backrestVolume, dataVolume, walVolume operator.StorageResult,
+	tablespaceVolumes map[string]operator.StorageResult,
+	err error,
```
In the future, would be good to put the return values in a `struct` of sorts.

jkatz commented on operator/cluster/clusterlogic.go:
```diff
@@ -236,12 +234,38 @@ func DeleteCluster(clientset *kubernetes.Clientset, restclient *rest.RESTClient,
+		st = cluster.Spec.UserLabels[config.LABEL_SERVICE_TYPE]
+	}
+
+	serviceName := replica.Spec.ClusterName + "-replica"
```
preference for `fmt.Sprintf`

jkatz commented on operator/cluster/clusterlogic.go:
```diff
@@ -279,26 +296,18 @@ func Scale(clientset *kubernetes.Clientset, client *rest.RESTClient, replica *cr
 	tablespaceStorageTypeMap := operator.GetTablespaceStorageTypeMap(cluster.Spec.TablespaceMounts)
 
+	// combine supplemental groups from all volumes
+	var supplementalGroups []int64
+	{
+		supplementalGroups = append(supplementalGroups, dataVolume.SupplementalGroups...)
+		for _, v := range tablespaceVolumes {
+			supplementalGroups = append(supplementalGroups, v.SupplementalGroups...)
+		}
+	}
```
I see this pattern enough, perhaps we just turn this into a function?

cbandy replied:
I think the right way forward is to spread it out rather than consolidate it. e.g. I would like tablespace code to modify a pod spec for tablespaces.

jkatz replied:
Makes sense. Looking forward to more discussions

cbandy commented:
@jmckulk I added variables to the installers.
@jkatz I made those syntax changes.

I also noticed that `kubeapi.CreatePVC` and `kubeapi.DeletePVCs` were no called, so I removed them.
